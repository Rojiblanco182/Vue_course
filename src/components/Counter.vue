<template>
  <h2 test-id="title">{{ getTitle }}</h2>
  <p> {{ counter }} <sup>2</sup> = {{squareCounter}} </p>
  <p data-testid="counter"> {{ counter }} </p>
  <div>
      <button v-on:click="increaseCounter" test-id="increase">+1</button>
      <button @click="decreaseCounter" test-id="decrease">-1</button>
  </div>
</template>

<script>
export default {
    props: {
        title: String,
        start: {
            type: Number,
            //required: true,
            default: 1,
            validator(value) {
                return value > 0
            }
        }
    },
    data() {
        return {
            counter: this.start
        }
    },
    methods: {
        increaseCounter() {
            this.counter++
        },
        decreaseCounter() {
            this.counter--
        }
    },
    computed: {
        squareCounter() {
            return this.counter * this.counter
        },
        getTitle() {
            return this.title || 'Counter'
        }
    }
}
</script>

<style>

button {
    margin: 0 5px;
}

</style>