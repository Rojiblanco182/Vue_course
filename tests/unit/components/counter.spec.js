import Counter from '@/components/Counter'
import { shallowMount } from '@vue/test-utils'

describe('Counter Component', () => {

    let wrapper

    beforeEach(() => {
        wrapper = shallowMount(Counter)
    })

    test('must match the snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })

    test(`should contain the default value 'Counter'`, () => {
        expect(wrapper.find('h2').exists()).toBeTruthy()

        const h2Value = wrapper.find('h2').text()

        expect(h2Value).toBe('Counter')
    })

    test(`should contain '1' as default value in the 'counter' paragraph`, () => {
        const targetValue = wrapper.find('[data-testid="counter"]').text()

        expect(targetValue).toBe('1')
    })
    
    test('should add/subtract 1 to the counter value when the increase/decrease button is pressed', async () => {
        const increaseButton = wrapper.find('[test-id="increase"]')
        const decreaseButton = wrapper.find('[test-id="decrease"]')

        await increaseButton.trigger('click')
        await decreaseButton.trigger('click')
        await decreaseButton.trigger('click')

        const counterValue = wrapper.find('[data-testid="counter"]').text()

        expect(counterValue).toBe('0')
    })
    
    test('should initialize the counter with the default value', () => {
        const { start } = wrapper.props()
        const counterValue = wrapper.find('[data-testid="counter"]').text()

        expect(Number(counterValue)).toBe(start)
    })
    
    test(`should show the received prop 'title'`, () => {
        const mockTitle = 'Mock Title'
        const wrapper = shallowMount(Counter, {
            props: {
                title: mockTitle
            }
        })

        const titleShown = wrapper.find('[test-id="title"]').text()

        expect(titleShown).toBe(mockTitle)
    })
    
    
})
