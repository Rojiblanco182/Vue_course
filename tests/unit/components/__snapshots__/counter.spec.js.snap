// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Component must match the snapshot 1`] = `
<h2 test-id="title">Counter</h2>
<p>1 <sup>2</sup> = 1</p>
<p data-testid="counter">1</p>
<div><button test-id="increase">+1</button><button test-id="decrease">-1</button></div>
`;
